File: Year_III/SEM_2/CT - Compilation Techniques/LAB/Project Python/libs/token.py
from enum import Enum, auto


class TokenCode(Enum):
    # identifiers
    ID = auto()

    # keywords
    IF = auto()
    ELSE = auto()
    WHILE = auto()
    FOR = auto()
    RETURN = auto()
    BREAK = auto()
    CONTINUE = auto()

    # types
    VOID = auto()
    INT = auto()
    DOUBLE = auto()
    CHAR = auto()
    STRUCT = auto()

    # constants
    CT_INT = auto()
    CT_REAL = auto()
    CT_CHAR = auto()
    CT_STRING = auto()

    # delimiters
    COMMA = auto()
    COLON = auto()
    SEMICOLON = auto()
    LPAREN = auto()
    RPAREN = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    LBRACE = auto()
    RBRACE = auto()

    # operators
    ASSIGN = auto()
    ADD = auto()
    SUB = auto()
    MUL = auto()
    DIV = auto()
    MOD = auto()
    NOT = auto()
    AND = auto()
    OR = auto()
    LT = auto()
    GT = auto()
    LE = auto()
    GE = auto()
    EQ = auto()
    NE = auto()
    DOT = auto()

    # others
    SPACE = auto()
    SLCOMMENT = auto()
    MLCOMMENT = auto()
    END = auto()


class Token:
    def __init__(self, code: TokenCode, line: int, column: int):
        self.code = code
        self.line = line
        self.column = column

    def __str__(self):
        return f'{self.line}:{self.column} - {self.code.name}'

    def error(self, msg: str) -> Exception:
        raise Exception(f'{self}: {msg}')


class ValuedToken(Token):
    def __init__(self, code: TokenCode, line: int, column: int, value: str | int | float):
        super().__init__(code, line, column)
        self.value = value

    def __str__(self):
        return f'{self.line}:{self.column} - {self.code.name} {self.value}'

File: year-III/semester-2/CT - Compilation Techniques/LAB/Project C/lib/token.c
#include "token.h"

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

/// error handling
char *token_type_name[] = {
    // identifiers
    "ID",
    // keywords
    "IF",
    "ELSE",
    "WHILE",
    "FOR",
    "RETURN",
    "BREAK",
    "CONTINUE",
    // types
    "VOID",
    "INT",
    "DOUBLE",
    "CHAR",
    "STRUCT",
    // constants
    "CT_INT",
    "CT_REAL",
    "CT_CHAR",
    "CT_STRING",
    // delimiters
    "COMMA",
    "COLON",
    "SEMICOLON",
    "LPAREN",
    "RPAREN",
    "LBRACKET",
    "RBRACKET",
    "LBRACE",
    "RBRACE",
    // operators
    "ASSIGN",
    "ADD",
    "SUB",
    "MUL",
    "DIV",
    "MOD",
    "NOT",
    "AND",
    "OR",
    "LT",
    "GT",
    "LE",
    "GE",
    "EQ",
    "NE",
    "DOT",
    // others
    "SPACE",
    "SLCOMMENT",
    "MLCOMMENT",
    "END",
};

void error(const char *format, ...) {
    va_list args;
    va_start(args, format);

    fprintf(stderr, "Error: ");
    vfprintf(stderr, format, args);
    fprintf(stderr, "\n");

    va_end(args);
    exit(1);
}

void tokenError(Token *token, const char *format, ...) {
    va_list args;
    va_start(args, format);

    fprintf(stderr, "Error: ");
    vfprintf(stderr, format, args);
    fprintf(stderr, " at line %d\n", token->line);

    va_end(args);
    exit(1);
}

void printToken(Token* token) {
    printf("%s", token_type_name[token->type]);
        switch (token->type) {
            // identifier
            case ID:
                printf(" (%s)", token->str);
                break;

            // constants
            case CT_CHAR:
                printf(" (%c / %d)", token->integer, token->integer);
                break;

            case CT_STRING:
                printf(" (%s)", token->str);
                break;

            case CT_INT:
                printf(" (%d)", token->integer);
                break;

            case CT_REAL:
                printf(" (%f)", token->real);
                break;

            // others
            case SLCOMMENT:
                printf(" (%s)", token->str);
                break;

            case MLCOMMENT:
                printf(" (%s)", token->str);
                break;

            default:
                break;
        }
        printf("\n");
}